### PR TITLE
feat: recursive copying with --up

### PR DESCRIPTION
**Pull request by sgtrusty**

### CODE REVIEW

**Comment by netrules** [comment]
needed for https://github.com/sindresorhus/cpy-cli/pull/33

File: index.d.ts [hunk 1]
```diff
@@ -53,6 +53,13 @@ export interface Options extends Readonly<GlobOptions>, CpFileOptions {
 	*/
 	readonly flat?: boolean;
 
+	/**
+	Trim path from files being copied.
+
+	@default 0
+	*/
+	readonly up?: boolean;
```
Comment (sindresorhus) [inline index.d.ts:61]: I don't think this is a good option name. Maybe `trimPathComponents` or something.

File: index.js [hunk 2]
```diff
@@ -74,17 +75,34 @@ const preprocessDestinationPath = ({entry, destination, options}) => {
...
+  if (!up) {
+    return source;
+  }
+  if (up === true) {
+    return path.basename(source);
+  }
+  if ((path.normalize(source).split(path.sep).length - 1) < up) {
+    throw new Error('cant go up that far');
+  }
+  return path.join(path.normalize(source).split(path.sep).slice(up));
```
Comment (sindresorhus) [inline index.js:103]: Use tab-indentation.

File: index.js [hunk 3]
```diff
@@ -74,17 +75,34 @@ const preprocessDestinationPath = ({entry, destination, options}) => {
...
 
+/**
+@param {string} source
+@param {boolean|int} up
+*/
+function trimPath(source, up) {
+  if (!up) {
+    return source;
+  }
+  if (up === true) {
```
Comment (sindresorhus) [inline index.js:97]: This case is not documented.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
This will need tests.

**Comment by sindresorhus** [comment]
Note that things have changed since you started this: https://github.com/sindresorhus/cpy/releases/tag/v9.0.0

**Review by tornadia — APPROVED** [review]
@sindresorhus please see changes commited

File: index.d.ts [hunk 4]
```diff
@@ -53,6 +53,13 @@ export interface Options extends Readonly<GlobOptions>, CpFileOptions {
 	*/
 	readonly flat?: boolean;
 
+	/**
+	Trim path from files being copied.
+
+	@default 0
```
Comment (sindresorhus) [inline index.d.ts:59]: Missing the example from the readme.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sgtrusty** [comment]
Sorry, been a bit busy. Gonna check em out and jump on it, thanks @sindresorhus

File: readme.md [hunk 5]
```diff
@@ -83,6 +83,24 @@ Default: `true`
...
+```js
+import cpy from 'cpy';
+
+await cpy('src/**/*.js', 'destination/flatSourceFiles', {
+	trimPathComponents: 2
+});
+```
+
+Where the destination paths of `*.js` files within subdirectories of the `src` directory
+will be flattened to a reduced structure another `destination/flatSourceFiles` path.
```
Comment (sindresorhus) [inline readme.md:102]: Don't hard-wrap.

This is missing from index.d.ts

And I think it needs a rewrite. I don't really understand what it's trying to say.

**Pull request closed without merging**